src/java.base/share/classes/java/lang/foreign/SymbolLookup.java

```diff
@@ -0,0 +1,242 @@
+ * through JNI (the association with a class loader is immaterial to the library lookup):
+ *
+ * {@snippet lang=java :
+ * System.loadLibrary("GL"); // libGL.so loaded here
+ * libraryLookup("libGL.so", session).lookup("glGetString").isPresent(); // true
+ * }
+ *
+ * <p>
+ * Finally, each {@link Linker} provides a symbol lookup for libraries that are commonly used on the OS and processor
+ * combination supported by that {@link Linker}. This symbol lookup, which is known as a <em>default lookup</em>,
+ * helps clients to quickly find addresses of well-known symbols. For example, a {@link Linker} for Linux/x64 might choose to
+ * expose symbols in {@code libc} through the default lookup:
+ *
+ * {@snippet lang=java :
+ * Linker nativeLinker = Linker.nativeLinker();
+ * SymbolLookup stdlib = nativeLinker.defaultLookup();
+ * MemorySegment malloc = stdlib.lookup("malloc");
+ * }
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
+@FunctionalInterface
+public interface SymbolLookup {
+
+    /**
+     * Returns the address of the symbol with the given name.
+     * @param name the symbol name.
+     * @return a zero-length memory segment whose base address indicates the address of the symbol, if found.
+     */
+    Optional<MemorySegment> lookup(String name);
+
+    /**
+     * Returns a symbol lookup for symbols in the libraries associated with the caller's class loader.
+     * <p>
+     * A library is associated with a class loader {@code CL} when the library is loaded via an invocation of
+     * {@link System#load(String)} or {@link System#loadLibrary(String)} from code in a class defined by {@code CL}.
+     * If that code makes further invocations of {@link System#load(String)} or {@link System#loadLibrary(String)},
+     * then more libraries are loaded and associated with {@code CL}. The symbol lookup returned by this method is always
+     * current: it reflects all the libraries associated with the relevant class loader, even if they were loaded after
+     * this method returned.
+     * <p>
+     * Libraries associated with a class loader are unloaded when the class loader becomes
+     * <a href="../../../java/lang/ref/package.html#reachability">unreachable</a>. The symbol lookup
+     * returned by this method is backed by a {@linkplain MemorySession#asNonCloseable() non-closeable}, shared memory
+     * session which keeps the caller's class loader reachable. Therefore, libraries associated with the caller's class
+     * loader are kept loaded (and their symbols available) as long as a loader lookup for that class loader is reachable.
+     *
+     * @return a symbol lookup for symbols in the libraries associated with the caller's class loader.
+     * @see System#load(String)
+     * @see System#loadLibrary(String)
+     */
+    @CallerSensitive
+    static SymbolLookup loaderLookup() {
+        Class<?> caller = Reflection.getCallerClass();
+        ClassLoader loader = Objects.requireNonNull(caller.getClassLoader());
```

Comment:
Good points. Regarding `ClassLoader` being null, I think we can still return something using the `BootLoader`'s `NativeLibraries` object - that would allow this method to be called internally. @mlchung Can you please confirm?

As for the caller sensitive having no real caller (e.g. because method is called directly from native code), I think we should add a check in `Reflection::ensureNativeAccess`. Few weeks ago I verified that adding such a check does not compromise upcall stub created via the Linker interface (e.g. a Java upcall might require to perform restricted operations - but those operation always happen inside the upcall MH, which is always associated with some class).

Reply:
Another option would be to treat calls to `ensureNativeAccess` with `null` caller class as coming from unnamed module.

Reply:
Looking deeper, `System::loadLibrary` seems to search the boot loader libraries when invoked with a `null` caller class, so replicating that behavior should be safe.

Reply:
`BootLoader` is what you want in this case - it defines the static methods to access resources, packages etc defined to the boot loader (aka null `ClassLoader`).

To find a symbol defined in the native libraries loaded by the boot loader, you can call `BootLoader.getNativeLibraries().find(name)`.

Reply:
When a caller-sensitive method is invoked from a thread attached via JNI, the caller class returned from `Reflection::getCallerClass` is `null`.    I would recommend the default to be a class in the unnamed module defined to the system class loader; hence it defaults to the system class loader.  This is consistent with JNI `FindClass` when invoked with no caller frame.

It's an open issue [1] to revisit the default for `System::load` and `System::loadLibrary` when invoked with null caller class.   However, the existing behavior may likely be unchanged because of the compatibility risk.

[1] https://bugs.openjdk.java.net/browse/JDK-8281001

Reply:
Thanks for the comments. I've pushed a new change which fixes a couple of thing:

* first, for `SymbolLookup.loaderLookup`, the system class loader is used when no caller class exists (e.g. when method is called from JNI). If caller class exist but its loader is null (boot loader), we just call ClassLoader::findNative with a `null` loader which will do the right thing (thanks @mlchung for the tips!)

* second, the restricted method check in `Reflection::ensureNativeAccess` has been improved to also work in case where caller class is `null`. In such cases, a dummy unnamed module module is used for the check.